README.md
# script.tvheadend.apiutils

A kodi addon to expose extra functionality from the Tvheadend backend, which isn't currently used by the pvr.hts addon.

Currently supports deleting all watched recordings, scheduling series links, and triggering an OTA EPG refresh.

After installing the addon, you must configure the Tvheadend server details in the addon settings.

You can then map the following actions to your kodi remote buttons (keyboard.xml / remote.xml), or include these in your skin:

* `RunScript(script.tvheadend.apiutils,delete-watched-recordings)`
   => This will only work from the TVRecordings window (WindowID: 10701) so must be placed in your <TVRecordings> section of the keymap.

* `RunScript(script.tvheadend.apiutils,toggle-series-link)`
   => This will only work from the TVGuide window (WindowID: 10702) so must be placed in your <TVGuide> section of the keymap.

* `RunScript(script.tvheadend.apiutils,refresh-epg)`
   => This will trigger an OTA EPG refresh in the Tvheadend backend. Note: this currently only supports the OTA grabbers.


## Installation

Launch Kodi >> Add-ons >> Get More >> .. >> Install from zip file

Enjoy!


script.tvheadend.apiutils/changelog.txt
2.0.1
- Initial Matrix release with Python 3.x compatability

1.0.6
- Improved the matching logic for the 'Delete Watched Recordings' functionality to work with EPG running state configurations

1.0.5
- Fixed the 'Toggle Series Link' functionality for Tvheadend v4.4 onwards

1.0.4
- Added support for Kodi Leia
- Fixed the error messages that should display when the wrong number of results are returned from the TVheadend API

1.0.3
- Added the ability to trigger an OTA EPG refresh

1.0.2
- Correction for scheduling future events immediately prior to a DST change due to a kodi bug

1.0.1
- Removed obsolete dependency script.module.requests
- Added a configuration option for what action to take when series link information is unavailable
- Tidied up some of the strings

1.0.0
- Initial Release of the TVHeadend API Utilities Addon
- Added function for removing all recordings in the Tvheadend backend which are marked as watched in the kodi front-end
- Added function to create a proper series link from the currently selected item in the TV Guide


script.tvheadend.apiutils/addon.py
# -*- coding: utf-8 -*-

"""
addon.py
~~~~~~~~~~~~~~~

This script takes an argument for which action to take and calls the relevant module.

By James Hutchinson 2017
"""

from datetime import datetime
start = datetime.now()
from common import handleException, displayError, log, logInfo, getString, setAddonExecutingStatus, logExecutionTime

log('Script Started')
try:
  setAddonExecutingStatus(True)
except RuntimeError as e:
  logInfo(str(e))
except:
  handleException(getString(32009))
else:
  from sys import argv
  if len(argv) > 1:
    if argv[1] == 'delete-watched-recordings':
      try:
        import deletewatched
        deletewatched.main()
      except RuntimeError as e:
        displayError(str(e))
      except:
        handleException(getString(32004) + '\n' + getString(32006))
    elif argv[1] == 'toggle-series-link':
      try:
        import toggleserieslink
        toggleserieslink.main()
        logExecutionTime(start)
      except RuntimeError as e:
        displayError(str(e))
      except:
        handleException(getString(32005) + '\n' + getString(32006))
    elif argv[1] == 'refresh-epg':
      try:
        import otaepggrab
        otaepggrab.main()
      except RuntimeError as e:
        displayError(str(e))
      except:
        handleException(getString(32010) + '\n' + getString(32006))
    else:
      displayError(getString(32001) + ": " + argv[1] + "\n" + getString(32003) )
  else:
    displayError(getString(32002) + "\n" + getString(32003) )

setAddonExecutingStatus(False)


script.tvheadend.apiutils/common.py
# -*- coding: utf-8 -*-

"""
common.py
~~~~~~~~~~~~~~~

This module is used to declare various useful common variables and functions.

By James Hutchinson 2017
"""

import traceback
import collections
import xbmc
import xbmcgui
import xbmcaddon
from dateutil.parser import *
from time import time,mktime,strptime,strftime,gmtime
from datetime import datetime

addon = xbmcaddon.Addon()
addonDebug = addon.getSetting('debug')
addonId = addon.getAddonInfo('id')
addonName = addon.getLocalizedString(32000)

tvhServername = addon.getSetting('tvh_servername')
tvhPort = addon.getSetting('tvh_port')
tvhUsername = addon.getSetting('tvh_user')
tvhPass = addon.getSetting('tvh_pass')

seriesLinkManualRecType = addon.getSetting('serieslink_manual_rec_type')
MANUAL_RECTYPE_KODI_PROMPT = '0'  # Display Kodi Timer Rule Prompt
MANUAL_RECTYPE_TVH_AUTOREC = '1'  # Create basic AutoRec entry via the TVHeadend API
MANUAL_RECTYPE_NO_ACTION   = '2'  # Display a Popup and Take No Action

kodiLongdateFormat = xbmc.getRegion('datelong')
kodiTimeFormatNoSeconds = xbmc.getRegion('time').replace(':%S','')
kodiLongDateTimeFormatNoSeconds = kodiLongdateFormat + ' ' + kodiTimeFormatNoSeconds

def setAddonExecutingStatus(newState):
  if newState == True:
    if xbmc.getInfoLabel('Window(10000).Property(script.tvheadend.apiutils.isrunning)') == 'True':
      raise RuntimeError('Another instance of the script is already running')
    else:
      xbmcgui.Window(10000).setProperty('script.tvheadend.apiutils.isrunning', 'True')
  else:
    xbmcgui.Window(10000).setProperty('script.tvheadend.apiutils.isrunning', 'False')

def getString(id):
  return addon.getLocalizedString(id)

def getKodiString(id):
  return xbmc.getLocalizedString(id)

def log(msg, level=xbmc.LOGDEBUG):
  if addonDebug == 'true' and level == xbmc.LOGDEBUG:
    level = xbmc.LOGINFO
  try:
    xbmc.log('[%s] %s' % (addonId, msg), level)
  except Exception as e:
    try:
      xbmc.log('Logging Failure: %s' % (e), level)
    except:
      pass # just give up

def logInfo(msg):
  log(msg, xbmc.LOGINFO)

def logError(errMsg):
  log(errMsg, xbmc.LOGWARNING)

def displayError(errMsg):
  logError(errMsg)
  notificationPopup(errMsg, 'DefaultIconError.png', addonName)

def handleException(onScreenMsg, logMsg='An unexpected error occurred'):
  displayError(onScreenMsg)
  logError(traceback.format_exc())

def notificationPopup(msg, icon='default', title=getKodiString(19166), timeout='5000'):
  args=title+','+msg+','+str(timeout)
  if icon != 'default': args=args+','+icon
  xbmc.executebuiltin('Notification(' + args + ')')

def okDialog(message, title=addonName):
  xbmcgui.Dialog().ok(title, message)

def yesNoDialog(message, title=addonName):
  return xbmcgui.Dialog().yesno(title, message)

def epochFromUTCTimestamp(strTimestamp):
  return int((parse(strTimestamp+'Z') - parse('1970-01-01 00:00:00Z')).total_seconds())

def epochFromLocalTimestamp(strTimestamp, strTimestampFormat=kodiLongDateTimeFormatNoSeconds):
  localDateTime = datetime.fromtimestamp(mktime(strptime(strTimestamp, strTimestampFormat)))
  utcOffset = datetime.now() - datetime.utcnow()
  utcDateTime = localDateTime - utcOffset
  epoch = datetime.utcfromtimestamp(0)
  return int((utcDateTime - epoch).total_seconds())

def epochNow():
  return int(time())

def convertDictToText(data):
  if isinstance(data, basestring):
    return str(data.encode('utf8'))
  elif isinstance(data, collections.Mapping):
    return dict(map(convertDictToText, data.iteritems()))
  elif isinstance(data, collections.Iterable):
    return type(data)(map(convertDictToText, data))
  else:
    return data

def getKodiInfoLabel(label, errorString=getString(32007)):
  retVal = xbmc.getInfoLabel(label)
  if not retVal:
    errorString = errorString + label
    raise RuntimeError(errorString)
  else:
    log(label + ': ' + retVal)
    return retVal

def getKodiCondVisibility(label, errorString=getString(32008)):
  retVal = xbmc.getCondVisibility(label)
  log(label + ': ' + str(retVal))
  return retVal

def getCurrentWindowId():
  return xbmcgui.getCurrentWindowId()

def xbmcExecuteBuiltin(action):
  xbmc.executebuiltin(action)

def logExecutionTime(start):
  logInfo('Script execution time: ' + str((datetime.now() - start).total_seconds()) + 's')

def kodiMajorVersion():
  return int(xbmc.getInfoLabel('System.BuildVersion')[:2])


script.tvheadend.apiutils/tvhapi.py
# -*- coding: utf-8 -*-

"""
tvhapi.py
~~~~~~~~~~~~~~~

This module is used to make various calls to the Tvheadend API.

By James Hutchinson 2017
"""

from common import log, logError, getString, tvhServername, tvhPort, tvhUsername, tvhPass
import urllib
import urllib.error
import urllib.parse
import urllib.request
import json

def raiseHTTPError(errMsg):
  logError(errMsg)
  onScreenErrMsg = getString(32300) + ' (' + 'http://' + tvhServername + ':' + tvhPort + ')'
  raise RuntimeError(onScreenErrMsg)

def httpAuth(url, authType='digest'):
  passwordMgr = urllib.request.HTTPPasswordMgrWithDefaultRealm()
  passwordMgr.add_password(None, url, tvhUsername, tvhPass)

  if authType == 'digest':
    log('Attempting Tvheadend Digest Authentication')
    authHandler = urllib.request.HTTPDigestAuthHandler(passwordMgr)
  else:
    log('Attempting Tvheadend Basic Authentication')
    authHandler = urllib.request.HTTPBasicAuthHandler(passwordMgr)
  return urllib.request.build_opener(authHandler)

def httpPost(url, encodedParams, authType='digest'):
  try:
    request = httpAuth(url, authType)
    response = request.open(url, data=encodedParams)
  except urllib.error.HTTPError as e:
    if e.code == 401:
      if authType == 'digest': #attempt to fallback to basic http auth
        log('Digest Authentication Failed')
        return httpPost(url, encodedParams, 'basic')
      else:
        logError(getString(32301))
    raiseHTTPError('HTTP error %s: %s' % (e.code, e.reason))
  except urllib.error.URLError as e:
    raiseHTTPError('URL error: %s: %s' % (e.code, e.reason))
  except Exception as e:
    raiseHTTPError('HTTP general exception: %s' % e)
  else:
    log('Tvheadend API Request Successful:')
    log('  URL: ' + url)
    log('  Parameters: ' + str(encodedParams))
    log('Content Type: '+ response.headers['content-type'])
    encoding = response.headers['content-type'].split('charset=')[-1]
    log('Encoding: ' + encoding)
    rawResponse = response.read().decode(encoding).encode('utf-8')
    log('Response: ' + str(rawResponse))
    return rawResponse

def tvhApiResponse(urlPath, urlParams):
  url = 'http://' + tvhServername + ':' + tvhPort + urlPath
  encodedParams = urllib.parse.urlencode(urlParams).encode('utf-8')
  response = httpPost(url, encodedParams)
  log('Converting JSON response')
  jsonResponse = json.loads(response, strict=False)
  log('JSON Conversion Complete')
  return jsonResponse

def getTvhUpcomingRecording(uuid, fields):
  return tvhApiResponse('/api/idnode/load', {'uuid':uuid,'list':fields})

def getTvhFinishedRecordings():
  return tvhApiResponse('/api/dvr/entry/grid_finished', {'limit':'999999999'})

def getTvhChannels():
  return tvhApiResponse('/api/channel/list', {'all':'1'})

def getTvhChannelEpg(channelId):
  return tvhApiResponse('/api/epg/events/grid', {'channel':channelId})

def getTvhEpgEvent(channelNumber, epochStartDateTime, epochEndDateTime):
  filterParams = '[{\"type\":\"numeric\",\"comparison\":\"eq\",\"value\":' + channelNumber + ',\"field\":\"channelNumber\"},' + \
                  '{\"type\":\"numeric\",\"value\":' + str(epochStartDateTime) + ',\"field\":\"start\"},' + \
                  '{\"type\":\"numeric\",\"value\":' + str(epochEndDateTime) + ',\"field\":\"stop\"}]'
  return tvhApiResponse('/api/epg/events/grid', {'filter':filterParams})

def deleteTvhFinishedRecording(uuid):
  return tvhApiResponse('/api/dvr/entry/remove', {'uuid':'[\"' + uuid + '\"]'})

def addTvhSeriesLink(eventId):
  return tvhApiResponse('/api/dvr/autorec/create_by_series', {'event_id':eventId,'config_uuid':''})

def addTvhAutoRecRule(tvhShowTitle, tvhChannelUuid, comment=getString(32302)):
  return tvhApiResponse('/api/dvr/autorec/create', {'conf':'{\"enabled\":1,\"comment\":\"' + comment + '\",\"title\":\"' + tvhShowTitle + '\",\"channel\":\"' + tvhChannelUuid + '\"}'})

def renameTvhIdNode(uuid, newName, newComment=getString(32302)):
  return tvhApiResponse('/api/idnode/save', {'node':'[{\"name\":\"' + newName + '\",\"comment\":\"' + newComment + '\",\"uuid\":\"' + uuid + '\"}]'})

def deleteTvhIdNode(uuid):
  return tvhApiResponse('/api/idnode/delete', {'uuid':'[\"' + uuid + '\"]'})

def triggerOtaEpgGrabber():
  return tvhApiResponse('/api/epggrab/ota/trigger', {'trigger':'1'})


script.tvheadend.apiutils/otaepggrab.py
# -*- coding: utf-8 -*-

"""
otaepggrab.py
~~~~~~~~~~~~~~~~~~~

This module is used to manually trigger the OTA EPG grabber via the Tvheadend API

By James Hutchinson 2018
"""

def main():

  from common import logInfo, logError, notificationPopup, getString

  import tvhapi

  def triggerTvhOtaEpgGrabber():
    logInfo('Triggering the OTA EPG grabber via the Tvheadend API')
    if tvhapi.triggerOtaEpgGrabber(): #A success yields a blank {} response
      logError('Encountered an invalid response when attempting to trigger the OTA EPG grabber via the Tvheadend API')
      raise RuntimeError(getString(32500))
    else:
      logInfo('Successfully triggered the OTA EPG grabber via the Tvheadend API')
      notificationPopup(getString(32501))

  #End of function declarations

  triggerTvhOtaEpgGrabber()


script.tvheadend.apiutils/toggleserieslink.py
# -*- coding: utf-8 -*-

"""
toggleserieslink.py
~~~~~~~~~~~~~~~~~~~

This module is used to toggle a series link timer from within the TV guide

By James Hutchinson 2017
"""

def main():

  from common import log, logInfo, logError, epochFromUTCTimestamp, epochFromLocalTimestamp, epochNow, getString, kodiMajorVersion
  from common import getKodiInfoLabel, getKodiCondVisibility, getCurrentWindowId, xbmcExecuteBuiltin, seriesLinkManualRecType
  from common import MANUAL_RECTYPE_KODI_PROMPT, MANUAL_RECTYPE_TVH_AUTOREC, MANUAL_RECTYPE_NO_ACTION

  import tvhapi

  def getCurrentEventEpochStartDateTime():
    startDateTimeStr = getKodiInfoLabel('ListItem.StartDate') + ' ' + getKodiInfoLabel('ListItem.StartTime')
    epochStartDateTime = epochFromLocalTimestamp(startDateTimeStr)
    log('Selected Event Start Date/Time (Epoch): ' + str(epochStartDateTime) )
    return epochStartDateTime

  def getCurrentEventEpochEndDateTime():
    endDateTimeStr = getKodiInfoLabel('ListItem.EndDate') + ' ' + getKodiInfoLabel('ListItem.EndTime')
    epochEndDateTime = epochFromLocalTimestamp(endDateTimeStr)
    log('Selected Event End Date/Time (Epoch): ' + str(epochEndDateTime) )
    return epochEndDateTime

  def getTvhAutorecUuid(tvhDvrUuid):
    tvhRecording = tvhapi.getTvhUpcomingRecording(tvhDvrUuid, 'autorec')["entries"]
    if len(tvhRecording) == 1:
      if tvhRecording[0]["params"][0]["id"] == 'autorec':
        autoRec = tvhRecording[0]["params"][0]["value"]
        if len(autoRec) == 32:
          return autoRec
        else:
          log('The Tvheadend Autorec lookup has returned an Invalid UUID')
          raise RuntimeError(getString(32207))
      else:
        log('Unexpected parameters were returned for the Tvheadend Autorec lookup')
        raise RuntimeError(getString(32207))
    else:
      log('The Tvheadend lookup found ' + str(len(tvhRecording)) + ' upcoming recordings for the current event, but we expected 1')
      raise RuntimeError(getString(32207))

  def getTvhEpgEvent(channelNumber, channelName, epochStartDateTime, epochEndDateTime):
    tvhChannelEpg = tvhapi.getTvhEpgEvent(channelNumber, epochStartDateTime, epochEndDateTime)
    tvhEpgEvent = [x for x in tvhChannelEpg["entries"] if x["channelName"] == channelName] #Check the channel name matches
    if len(tvhEpgEvent) == 1: #Make sure that one and only one event was found
      thisEvent = tvhEpgEvent[0]
      if "dvrUuid" in thisEvent:
        thisEvent.setdefault("autorec", getTvhAutorecUuid(thisEvent["dvrUuid"]))
      else:
        thisEvent.setdefault("autorec", 0)
      return thisEvent
    else:
      log('The Tvheadend lookup found ' + str(len(tvhEpgEvent)) + ' events, but we expected 1')
      raise RuntimeError(getString(32202))

  def tvhEventHasSeriesLinkInfo(tvhEpgEvent):
    if "serieslinkUri" in tvhEpgEvent:
      logInfo('Series Link information was found for this event')
      logInfo('Series Link URI: ' + str(tvhEpgEvent["serieslinkUri"]))
      return True
    else:
      logInfo('Series Link information was not availble for this event')
      return False

  def renameTvhTimer(tvhTimerUuid, tvhShowTitle, tvhTimerComment):
    log('UUID found in timer rule creation response from Tvheadend: ' + str(tvhTimerUuid))
    if len(tvhTimerUuid) == 32:
      log('Renaming the timer rule in the Tvheadend backend')
      tvhRenameNodeResponse = tvhapi.renameTvhIdNode(tvhTimerUuid, tvhShowTitle, tvhTimerComment) #A success yields a blank {} response
      if tvhRenameNodeResponse:
        log('Encountered an invalid response when attempting to rename the Tvheadend timer rule')
        raise RuntimeError(getString(32206))
    else:
      log('Encountered an invalid UUID when attempting to rename the Tvheadend timer rule')
      raise RuntimeError(getString(32206))

  def createTvhSeriesLink(tvhEpgEvent):
    logInfo('Creating series link via the Tvheadend API for: ' + tvhEpgEvent["title"])
    try:
      seriesLinkResponse = tvhapi.addTvhSeriesLink(tvhEpgEvent["eventId"])
      tvhTimerUuid = seriesLinkResponse["uuid"][0]
      tvhTimerComment = getString(32208) + ' URI: ' + tvhEpgEvent["serieslinkUri"]
      logInfo('Series link successfully scheduled via the Tvheadend API')
      renameTvhTimer(tvhTimerUuid, tvhEpgEvent["title"], tvhTimerComment)
    except:
      logError('Invalid series link response from Tvheadend')
      raise RuntimeError(getString(32204))

  def showKodiTimerRulePrompt():
    logInfo('Displaying the kodi built-in create timer rule window')
    xbmcExecuteBuiltin('Action(ShowTimerRule)')

  def createTvhAutoRec(tvhEpgEvent):
    logInfo('Creating timer rule via the Tvheadend API for: ' + tvhEpgEvent["title"])
    try:
      timerRuleResponse = tvhapi.addTvhAutoRecRule(tvhEpgEvent["title"], tvhEpgEvent["channelUuid"])
      tvhTimerUuid = timerRuleResponse["uuid"]
      tvhTimerComment = getString(32209)
      logInfo('Timer tule successfully scheduled via the Tvheadend API')
      renameTvhTimer(tvhTimerUuid, tvhEpgEvent["title"], tvhTimerComment)
    except:
      logError('Invalid timer rule creation response from Tvheadend')
      raise RuntimeError(getString(32205))

  def createTimer(tvhEpgEvent):
    if tvhEventHasSeriesLinkInfo(tvhEpgEvent):
      createTvhSeriesLink(tvhEpgEvent)
    else:
      if seriesLinkManualRecType == MANUAL_RECTYPE_KODI_PROMPT:
        showKodiTimerRulePrompt()
      elif seriesLinkManualRecType == MANUAL_RECTYPE_TVH_AUTOREC:
        createTvhAutoRec(tvhEpgEvent)
      else:
        raise RuntimeError(getString(32210))

  def deleteTimer(tvhEpgEvent):
    logInfo('Removing timer rule via the Tvheadend API for: ' + tvhEpgEvent["title"])
    autorecUuid = tvhEpgEvent["autorec"]
    log('Autorec UUID: ' + autorecUuid)
    if tvhapi.deleteTvhIdNode(autorecUuid): #A success yields a blank {} response
      logError('Encountered an invalid response when attempting to delete the Tvheadend timer rule')
      raise RuntimeError(getString(32203))

  def checkWindowId():
    log('WindowID: ' + str(getCurrentWindowId()) )
    if getCurrentWindowId() != 10702:
      raise RuntimeError(getString(32200))

  def checkEpgEventStatus():
    if getKodiCondVisibility('ListItem.IsRecording'):
      raise RuntimeError(getString(32201))

  def getKodiChannelNumber():
    if kodiMajorVersion() < 18:
      return getKodiInfoLabel('ListItem.ChannelNumber')
    else:
      return getKodiInfoLabel('ListItem.ChannelNumberLabel')

  def toggleSeriesLink():
    logInfo('Attempting to toggle series link for: ' + getKodiInfoLabel('ListItem.EPGEventTitle'))

    epochStartDateTime = getCurrentEventEpochStartDateTime()
    epochEndDateTime = getCurrentEventEpochEndDateTime()

    if epochEndDateTime > epochNow():
      tvhEpgEvent = getTvhEpgEvent(getKodiChannelNumber(), \
                                   getKodiInfoLabel('ListItem.ChannelName'), \
                                   epochStartDateTime, \
                                   epochEndDateTime )
      if tvhEpgEvent["autorec"]: #If a timer rule has already been setup then remove it
        deleteTimer(tvhEpgEvent)
      else:
        createTimer(tvhEpgEvent)
    else:
      logInfo('Event is in the past')
      showKodiTimerRulePrompt()

  #End of function declarations

  checkWindowId()
  checkEpgEventStatus()
  toggleSeriesLink()


script.tvheadend.apiutils/deletewatched.py
# -*- coding: utf-8 -*-

"""
deletewatched.py
~~~~~~~~~~~~~~~

This module is used to delete all recordings from tvheadend that are marked watched in kodi

By James Hutchinson 2017
"""

def main():

  from common import log, logInfo, logError, epochFromUTCTimestamp, getString, notificationPopup, getCurrentWindowId, okDialog, yesNoDialog, kodiMajorVersion
  import kodiapi
  import tvhapi

  def filterWatchedKodiRecordings(kodiRecordings):
    return [x for x in kodiRecordings["result"]["recordings"] if x["playcount"] > 0]

  def deletionConfirmMsg(numRecordings):
    if numRecordings == 1:
      return getString(32103) + '\n' + getString(32106)
    else:
      return getString(32104) + ' ' + str(numRecordings) + ' ' + getString(32105) + '\n' + getString(32106)

  def getTvhRecordingUuid(kodiWatchedRecording, tvhRecordings):
    kodiEpochStartTime = epochFromUTCTimestamp(kodiWatchedRecording["starttime"])
    logInfo('Found Watched Recording:')
    logInfo('  Channel: ' + kodiWatchedRecording["channel"])
    logInfo('  Title: ' + kodiWatchedRecording["title"])
    logInfo('  Start Date/Time: ' + kodiWatchedRecording["starttime"])
    logInfo('  Epoch Start Time: ' + str(kodiEpochStartTime))
    logInfo('  End Date/Time: ' + kodiWatchedRecording["endtime"])
    logInfo('  Plot: ' + kodiWatchedRecording["plot"])

    tvhRecording = [x for x in tvhRecordings["entries"] \
                        if     kodiWatchedRecording["channel"] == x["channelname"] \
                       and     kodiWatchedRecording["title"]   == x["disp_title"] \
                       and     kodiEpochStartTime >= x["start_real"] \
                       and     kodiEpochStartTime < x["stop_real"] \
                       and (   kodiWatchedRecording["plot"] == x["disp_description"] \
                            or kodiWatchedRecording["plot"] == x["disp_subtitle"] ) ]
    if len(tvhRecording) == 1: #Make sure that one and only one recording was found
      tvhRecordingUuid = str(tvhRecording[0]["uuid"])
      logInfo('  Tvheadend Recording UUID: ' + tvhRecordingUuid)
      return tvhRecordingUuid
    else:
      logError('The Tvheadend lookup found ' + str(len(tvhRecording)) + ' recordings, but we expected 1')
      return 0

  def deleteWatchedRecording(kodiWatchedRecording, tvhRecordings):
    success = False

    tvhRecordingUuid = getTvhRecordingUuid(kodiWatchedRecording, tvhRecordings)
    if tvhRecordingUuid:
      if len(tvhRecordingUuid) == 32:
        logInfo('Deleting Watched Recording UUID: ' + tvhRecordingUuid)
        deletionResponse = tvhapi.deleteTvhFinishedRecording(tvhRecordingUuid)
        if not deletionResponse: #Successful deletion yields a blank {} response
          logInfo('Recording Successfully Deleted')
          success = True
        else:
          logError('Deletion failed, invalid response from the Tvheadend API')
      else:
        logError('Skipping deletion since the Tvheadend lookup returned an invalid UUID')
    else:
      logError('Skipping deletion, since the Tvheadend lookup failed to identify a unique entry')
    logInfo('--------------------------------------------------------------------------------')
    return success

  def checkWindowId():
    log('WindowID: ' + str(getCurrentWindowId()) )
    if getCurrentWindowId() != 10701:
      raise RuntimeError(getString(32100))

  def deleteAllWatchedRecordings():
    logInfo('Attempting to delete all watched recordings')
    logInfo('Gathering list of recordings from Kodi')
    kodiRecordings = kodiapi.getRecordings()
    logInfo('Filtering out watched Kodi recordings')
    kodiWatchedRecordings = filterWatchedKodiRecordings(kodiRecordings)

    if len(kodiWatchedRecordings) == 0:
      logInfo('No watched recordings were found')
      okDialog(getString(32102), getString(32101))
    else:
      logInfo(str(len(kodiWatchedRecordings)) + ' watched recordings were found')
      logInfo('Prompting for confirmation before proceeding')
      confirmMsg = deletionConfirmMsg(len(kodiWatchedRecordings))
      if yesNoDialog(confirmMsg, getString(32101)):
        logInfo('User answered yes to the confirmation prompt, so continuing')
        logInfo('Gathering list of recordings from Tvheadend')
        tvhRecordings = tvhapi.getTvhFinishedRecordings()

        numFilesDeleted = 0
        logInfo('----------------------------------------------------------------------')
        for kodiWatchedRecording in kodiWatchedRecordings:
          if deleteWatchedRecording(kodiWatchedRecording, tvhRecordings): numFilesDeleted += 1

        notificationPopup(str(numFilesDeleted) + ' ' + getString(32107))
        logInfo(str(numFilesDeleted) + ' watched recording(s) were deleted')
      else:
        logInfo('The user answered no to the confirmation prompt, so skipping deletion')

  #End of function declarations
  checkWindowId()
  deleteAllWatchedRecordings()
